src/MongoDB.Driver.Core/Core/Authentication/SaslPrepHelper.cs

```diff
@@ -0,0 +1,806 @@
+﻿/* Original work:
+ *   Copyright 2017 Tom Bentley
+ * 
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *   
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *   
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ * 
+ * Modified work: 
+ *   Copyright 2018–present MongoDB Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *   
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *   
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// Utility class for Sasl string preparation.
+    /// </summary>
+    internal static class SaslPrepHelper
+    {
+        internal const int MinCodepoint = 0x00;
+        internal const int MaxCodepoint = 0x10FFFF;
+        internal const int SurrogateMinCodepoint = 0x00d800;
+        internal const int SurrogateMaxCodePoint = 0x00dfff;
+        
+        /// <summary>
+        /// Return the SASLPrep-canonicalised version of the given <paramref name="str"/> for use as a query string.                          
+        /// This implements the {@code SASLPrep} algorithm defined in <a href="https://tools.ietf.org/html/rfc4013">RFC 4013</a>.          
+        /// See <a href="https://tools.ietf.org/html/rfc3454#section-7">RFC 3454, Section 7</a> for discussion of what a
+        /// query string is.
+        /// String normalization step in the .NET Standard version of the driver is skipped due to a lack of a string
+        /// normalization function.
+        /// </summary>
+        /// <param name="str">The string to canonicalise.</param>
+        /// <returns>The canonicalised string.</returns>
+        internal static string SaslPrepQuery(string str)
```

Comment:
The class itself is internal. All the internal methods can be public. (But the private ones should stay private).

Reply:
Fixed!